_episodes/13.md

```diff
@@ -0,0 +1,449 @@
+questions based on a specific customer, or even a specific product. The
+questions themselves are typically have answers involving aggregate data like
+sums, averages, counts, etc... Wouldn't it make sense to cache some of these
+results? 
+
+A common way to visualize the columns that are commonly bucketed to some values
+or range of values is to show them as a cube, that is sliced up into smaller
+dimensions. This actually derives from the traditional form of OLAP, 
+multi-dimensional OLAP (MOLAP). 
+
+<p align="center">
+<img align="center" width="75%" height="100%" src="/assets/episode/13/data_cube.png"/><br/>
+</p>
+
+This cube represents a caching of data aggregations that are grouped by commonly
+used dimensions. For example, the above cube would be the pre-aggregation of the
+following query:
+
+```
+SELECT part, store, customer, COUNT(*)
+FROM cube_table
+GROUP BY part, store, customer
+```
+
+If we want to get the data for a particular customer, we can take a "slice" of
+that cube by specifying a particular customer. The following query would give 
+us the green square above from our cube.
+
+```
+SELECT part, store, COUNT(*)
+FROM cube_table
+WHERE customer = "Bob"
+GROUP BY part, store
+```
+
+Now what if we want to flatten one of the dimensions? While this can be managed
+with a GROUP BY as before, but depending on the system may ignore any cached
+data and scan over all the rows. For this, SQL reserved a special set of
+keywords around cubes. We won't dive into that in depth now but for our current
+goal of flattening a dimension, we can use ROLLUP. Using the keyword ROLLUP
+indicates to the underlying system that you intend to aggregate over the 
+pre-materialized data rather than scan over all rows to compute again. This will
+give you the total count of parts per store using the counts of the data cube.
+
+```
+SELECT part, store, COUNT(*)
+FROM cube_table
+GROUP BY ROLLUP (part, store)
+```
+
+Now, although we used simple counts, you can precompute a lot of other aggregate
+data like sums, min, max, percentile, etc... These can service various queries
+that are commonly queried and doesn't require a new computation every time. That
+is the goal of MOLAP and data cubes.
+
+### Apache Pinot
+
+Now let's move on to Apache Pinot is a Realtime distributed OLAP datastore, 
+designed to answer OLAP queries with low latency. Although there may be a lot of
+words there that overlap with the Trino description, the key differentiators are
+realtime and low latency. Trino performs batch processing and is not a realtime
+system where Pinot is great for ingesting data in batch or stream. The other key
+word, low latency could technically apply to both Pinot and Trino but in the
+context of realtime subsecond latency, Trino is slow compared to Pinot. This
+is due to the specialized indexes that Pinot uses to store the data that we
+will cover shortly. Xiang has a really great summary slide that easily shows
```

Comment:
that we cover shortly. Also importantly Trino does not store any data itself. It is purely a query engine.